File: server/src/main/scala/com/netty/threed/ServerRequestHandler.scala
package com.netty.threed

import io.netty.buffer.ByteBuf
import io.netty.channel.{ChannelHandlerContext, ChannelInboundHandlerAdapter}
import io.netty.util.ReferenceCountUtil

class ServerRequestHandler extends ChannelInboundHandlerAdapter {

  override def channelRead(ctx: ChannelHandlerContext, event: scala.Any): Unit = {
    event match {
      case m: ByteBuf =>
        try {
          while (m.isReadable) {
            println(m.readByte())
          }
        } finally {
          ReferenceCountUtil.release(event)
        }
      case _ => println("oops")
    }
  }

  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
    cause.printStackTrace()
    ctx.close()
  }
}

File: server/src/main/java/com/netty/microservice/ServerRequestHandler.java
package com.netty.microservice;

import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelFutureListener;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;

public class ServerRequestHandler extends ChannelInboundHandlerAdapter {

    public void channelActive(ChannelHandlerContext ctx) {
        System.out.println("ChannelHandlerContext: " + ctx);
        var response = ctx.alloc().buffer(8); //64 bit int
        response.writeLong(System.currentTimeMillis());

        var channelTask = ctx.writeAndFlush(response);

        channelTask.addListener(new ChannelFutureListener() {
            public void operationComplete(ChannelFuture f) {
                ctx.close();
            }
        });

    }

    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        cause.printStackTrace();
        ctx.close();
    }

}
